math/const.go

```diff
@@ -22,12 +22,13 @@ const (
 )
 
 var (
-	_E     = newDecimal("2.718281828459045235360287471352662497757247093699959574966967627724076630353547594571382178525166427")
-	_Ln10  = newDecimal("2.302585092994045684017991454684364207601101488628772976033327900967572609677352480235997205089598298")
-	_Pi    = newDecimal("3.141592653589793238462643383279502884197169399375105820974944592307816406286208998628034825342117068")
-	_Pi2   = newDecimal("1.570796326794896619231321691639751442098584699687552910487472296153908203143104499314017412671058534")
-	_Sqrt3 = newDecimal("1.732050807568877293527446341505872366942805253810380628055806979451933016908800037081146186757248576")
-
+	_E                  = newDecimal("2.718281828459045235360287471352662497757247093699959574966967627724076630353547594571382178525166427")
+	_Ln10               = newDecimal("2.302585092994045684017991454684364207601101488628772976033327900967572609677352480235997205089598298")
+	_Pi                 = newDecimal("3.141592653589793238462643383279502884197169399375105820974944592307816406286208998628034825342117068")
+	_Pi2                = newDecimal("1.570796326794896619231321691639751442098584699687552910487472296153908203143104499314017412671058534")
+	_Sqrt3              = newDecimal("1.732050807568877293527446341505872366942805253810380628055806979451933016908800037081146186757248576")
+	_Pi_cache           *decimal.Big
```

Comment:
> This particular bit (_Pi_cache and _Pi_cache_precision) won't work because it's racy.

That is a good catch. I always assumed this was being run single threaded. If not single threaded then this is definitely a problem. And could definitely have a HUGE impact on result accuracy.

>  Do you have any data on the precision used in your programs?

For my programs yes, there could be multiple precisions used but at invocation time for decimal.XXX(ctx) the ctx would be the expected precision.
-----
I didn't realize the whole library is thread safe. That is awesome and something you should advertise in the README.md :-).

---
This could be mitigated by using a sync.RWMutex, but that would mean there would be a mux in there.  

It kind of sounds like you have an alternative.  What's your thoughts?

Reply:
> I didn't realize the whole library is thread safe. That is awesome and something you should advertise in the README.md :-).

Sorta! I outline this in the godocs:

> Unless otherwise specified, the only argument that will be modified is the result, typically a receiver named “z”. This means the following is valid and race-free:
> 
> x := New(1, 0)
> var z1, z2 Big
> 
> go func() { z1.Add(x, x) }()
> go func() { z2.Add(x, x) }()
> However, this is not:
> 
> x := New(1, 0)
> var z Big
> 
> go func() { z.Add(x, x) }() // BAD! RACE CONDITION!
> go func() { z.Add(x, x) }() // BAD! RACE CONDITION!

Any caches like github.com/ericlagergren/decimal/internal/arith/pow.go?s=1864:1896#L91 should be thread safe.

I'll expand more later today.

Reply:
Ok so after a bit of deliberation, I've decided I don't like any of the implementations using locks that I tried (maybe later if someone complains about it we can do something to help at that time).  So I have removed the Pi cache part. So i've only swapped out the continuing fractions form for piChudnovskyBrothers(). Since anything under 100 precision we use the const, this should provide some significant speedups for anyone using it for bigger precision sizes. 

BenchmarkPiContinuedFrac/5-16  	  500000	      **2301** ns/op
BenchmarkPiContinuedFrac/6-16  	  500000	      **2662** ns/op
BenchmarkPiContinuedFrac/7-16  	  300000	      **3432** ns/op
BenchmarkPiContinuedFrac/8-16  	  300000	      **3770** ns/op
BenchmarkPiContinuedFrac/9-16  	  300000	      **4667** ns/op
BenchmarkPiContinuedFrac/10-16 	  300000	      **5201** ns/op
BenchmarkPiContinuedFrac/11-16 	  200000	      **6829** ns/op
BenchmarkPiContinuedFrac/12-16 	  200000	      **7470** ns/op
BenchmarkPiContinuedFrac/13-16 	  200000	      8971 ns/op
BenchmarkPiContinuedFrac/14-16 	  200000	      9684 ns/op
BenchmarkPiContinuedFrac/15-16 	  200000	     11007 ns/op
BenchmarkPiContinuedFrac/16-16 	  100000	     12767 ns/op
BenchmarkPiContinuedFrac/25-16 	   30000	     40834 ns/op
BenchmarkPiContinuedFrac/32-16 	   30000	     55971 ns/op
BenchmarkPiContinuedFrac/41-16 	   20000	     79570 ns/op
BenchmarkPiContinuedFrac/50-16 	   10000	    104877 ns/op
BenchmarkPiContinuedFrac/75-16 	   10000	    180373 ns/op
BenchmarkPiContinuedFrac/82-16 	   10000	    202016 ns/op
BenchmarkPiContinuedFrac/97-16 	    5000	    269603 ns/op
BenchmarkPiContinuedFrac/100-16         	    5000	    266486 ns/op
BenchmarkPiContinuedFrac/137-16         	    5000	    385980 ns/op
BenchmarkPiContinuedFrac/250-16         	    2000	   1089872 ns/op
BenchmarkPiContinuedFrac/333-16         	    1000	   1795716 ns/op
BenchmarkPiContinuedFrac/500-16         	     500	   3574321 ns/op
BenchmarkPiContinuedFrac/646-16         	     200	   6868965 ns/op
BenchmarkPiContinuedFrac/750-16         	     200	   9044343 ns/op
BenchmarkPiContinuedFrac/943-16         	     100	  16091890 ns/op
BenchmarkPiContinuedFrac/1500-16        	      30	  36687125 ns/op
BenchmarkPiContinuedFrac/5000-16        	       2	 610959027 ns/op
BenchmarkPiContinuedFrac/7500-16        	       1	1788974484 ns/op
BenchmarkPiContinuedFrac/15000-16       	       1	11517741370 ns/op

BenchmarkPiChudnovskyBrothers/5-16  	  200000	      6597 ns/op
BenchmarkPiChudnovskyBrothers/6-16  	  200000	      7256 ns/op
BenchmarkPiChudnovskyBrothers/7-16  	  200000	      7145 ns/op
BenchmarkPiChudnovskyBrothers/8-16  	  200000	      6696 ns/op
BenchmarkPiChudnovskyBrothers/9-16  	  200000	      7202 ns/op
BenchmarkPiChudnovskyBrothers/10-16 	  200000	      6650 ns/op
BenchmarkPiChudnovskyBrothers/11-16 	  200000	      6952 ns/op
BenchmarkPiChudnovskyBrothers/12-16 	  200000	      8863 ns/op
BenchmarkPiChudnovskyBrothers/13-16 	  200000	      **8429** ns/op
BenchmarkPiChudnovskyBrothers/14-16 	  200000	      **9456** ns/op
BenchmarkPiChudnovskyBrothers/15-16 	  200000	     **10067** ns/op
BenchmarkPiChudnovskyBrothers/16-16 	  200000	      **9973** ns/op
BenchmarkPiChudnovskyBrothers/25-16 	  200000	      **9163** ns/op
BenchmarkPiChudnovskyBrothers/32-16 	  200000	     **11273** ns/op
BenchmarkPiChudnovskyBrothers/41-16 	  100000	     **12885** ns/op
BenchmarkPiChudnovskyBrothers/50-16 	  100000	     **13671** ns/op
BenchmarkPiChudnovskyBrothers/75-16 	  100000	     **18122** ns/op
BenchmarkPiChudnovskyBrothers/82-16 	  100000	     **19490** ns/op
BenchmarkPiChudnovskyBrothers/97-16 	  100000	     **22131** ns/op
BenchmarkPiChudnovskyBrothers/100-16         	  100000	     **21690** ns/op
BenchmarkPiChudnovskyBrothers/137-16         	   50000	     **26556** ns/op
BenchmarkPiChudnovskyBrothers/250-16         	   30000	     **55915** ns/op
BenchmarkPiChudnovskyBrothers/333-16         	   20000	     **65747** ns/op
BenchmarkPiChudnovskyBrothers/500-16         	   10000	    **107226** ns/op
BenchmarkPiChudnovskyBrothers/646-16         	   10000	    **140266** ns/op
BenchmarkPiChudnovskyBrothers/750-16         	   10000	    **158381** ns/op
BenchmarkPiChudnovskyBrothers/943-16         	   10000	    **219730** ns/op
BenchmarkPiChudnovskyBrothers/1500-16        	    5000	    **386462** ns/op
BenchmarkPiChudnovskyBrothers/5000-16        	    1000	   **1194042** ns/op
BenchmarkPiChudnovskyBrothers/7500-16        	    1000	   **1968074** ns/op
BenchmarkPiChudnovskyBrothers/15000-16       	     200	   **6001494** ns/op